File: 22. Generate Parentheses.java
class Solution {
    // public boolean check(String s){
    //     int score = 0;
    //     for(int i = 0; i < s.length(); i++){
    //         if(s.charAt(i) == '(') score++;
    //         else score--;

    //         if(score < 0) return false;
    //     }
    //     return true;
    // }
    public void backtrack(int open, int close, StringBuilder sb, int n, List<String> arr){
        if(sb.length() == n * 2){
            arr.add(sb.toString());
            return;
        }

        if(open < n){
            sb.append('(');
            backtrack(open + 1, close, sb, n, arr);
            sb.deleteCharAt(sb.length() - 1);
        }

        if(close < open){
            sb.append(")");
            backtrack(open, close + 1, sb, n, arr);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> arr = new ArrayList<>();
        backtrack(0, 0, new StringBuilder(), n, arr);

        return arr;
    }
}

File: 22. Generate Parentheses.cpp
class Solution {
public:
    bool check(const string& s) {
        int value = 0;
        for (int i = s.length() - 1; i >= 0; --i) {
            if (s[i] == ')') value++;
            else value--;

            if (value < 0) return false;
        }
        return true;
    }

    void exploration(vector<string>& res, string s, int leftPara, int rightPara) {
        if (leftPara == 0 && rightPara == 0) {
            if (check(s)) {
                res.push_back(s);
            }
            return;
        } 
        else if (leftPara == 0) {
            exploration(res, s + ")", leftPara, rightPara - 1);
        }
        else if (rightPara == 0) {
            exploration(res, s + "(", leftPara - 1, rightPara);
        }
        else {
            exploration(res, s + "(", leftPara - 1, rightPara);
            exploration(res, s + ")", leftPara, rightPara - 1);
        }
    }

    vector<string> generateParenthesis(int n) {
        vector<string> arr;
        exploration(arr, "(", n - 1, n);
        return arr;
    }
};
